keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1257 @@
+### Pod annotations
+
+Use Pod annotation as the initial K8s user interface, similar to e.g. how
+seccomp support was added. This will bridge the gap between enabling
+class-based resources in the CRI protocol and making them available in the Pod
+spec.
+
+A feature gate ClassResources enables kubelet to look for pod
+annotations and set the class resource assignment via CRI protocol accordingly.
+
+Specifically, kubelet will support annotations for specifying RDT and blockio
+class, the two types of class resources that already have basic support in the
+container runtimes.
+
+- `rdt.resources.beta.kubernetes.io/pod` for setting a Pod-level default RDT
+  class for all containers
+- `rdt.resources.beta.kubernetes.io/container.<container-name>` for
+  container-specific RDT class settings
+- `blockio.resources.beta.kubernetes.io/pod` for setting a Pod-level default
+  blockio class for all containers
+- `blockio.resources.beta.kubernetes.io/container.<container-name>` for
+  container-specific blockio class settings
+
+### Container runtimes
+
+We have implemented class-based RDT and blockio support in CRI-O and
+containerd:
+
+- cri-o:
+  - [~~Add support for Intel RDT~~](https://github.com/cri-o/cri-o/pull/4830)
+  - [~~Support for cgroups blockio~~](https://github.com/cri-o/cri-o/pull/4873)
+- containerd:
+  - [~~Support Intel RDT~~](https://github.com/containerd/containerd/pull/5439)
+  - [~~Support for cgroups blockio~~](https://github.com/containerd/containerd/pull/5490)
+
+The design paradigm here is that the container runtime configures the resource
+classes according to a given configuration file. Enforcement on containers is
+done via OCI. User interface is provided through pod and container annotations.
+
+### Open Questions
+
+#### Pod QoS class
+
+The Pod QoS class could be communicated to the container runtime as a class
+resource, too. This information is currently internal to kubelet. However,
+container runtimes (CRI-O, at least) are already depending on this information
+and currently determining it indirectly by evaluating other CRI parameters. It
+would be better to explicitly state the Pod QoS class and class resources would
+look like a logical place for that. This also makes it techically possible to
+have container-specific QoS classes (as a possible future enhancement of K8s).
+
+Communicating Pod QoS class via class resources would advocate moving class
+resources up to `ContainerConfig`.
+
+Making this change, it would also be possible to separate `oom_score_adj` from
+the pod qos class in the future.  The runtime could provide a set of OOM
+classes, making it possible for the user to specify a burstable pod with low
+oom priority (low chance of being killed).
+
+#### Default class
+
+A mechanism for indicating that the (runtime) default class should be used. The
+default class would/should be a node/runtime specific attribute. How should
+this be specified in the CRI protocol/`cri-api` and Pod spec?
+
+### Future work
```

Comment:
this section is extremely important for the KEP understanding. I'd suggest to put it on top after motivation. Same as we did in user namespaces KEP - outline the roadmap first and then stages

Reply:
K, got that. I moved "Future work" upwards, under a new "Implementation phases" section (copied that from #127).